crates/ironrdp-connector/src/channel_connection.rs

```diff
@@ -125,49 +126,46 @@ impl Sequence for ChannelConnectionSequence {
 
                 let user_channel_id = attach_user_confirm.initiator_id;
 
-                // user channel ID must also be joined
-                self.channel_ids.push(user_channel_id);
+                // User channel ID must also be joined.
+                self.channel_ids.insert(user_channel_id);
 
                 debug!(message = ?attach_user_confirm, user_channel_id, "Received");
 
                 debug_assert!(!self.channel_ids.is_empty());
 
                 (
                     Written::Nothing,
-                    ChannelConnectionState::SendChannelJoinRequest {
-                        user_channel_id,
-                        index: 0,
-                    },
+                    ChannelConnectionState::SendChannelJoinRequest { user_channel_id },
                 )
             }
 
-            // TODO(#112): send all the join requests in a single batch
-            // (RDP 4.0, 5.0, 5.1, 5.2, 6.0, 6.1, 7.0, 7.1, 8.0, 10.2, 10.3,
-            // 10.4, and 10.5 clients send a Channel Join Request to the server only after the
-            // Channel Join Confirm for a previously sent request has been received. RDP 8.1,
-            // 10.0, and 10.1 clients send all of the Channel Join Requests to the server in a
-            // single batch to minimize the overall connection sequence time.)
-            ChannelConnectionState::SendChannelJoinRequest { user_channel_id, index } => {
-                let channel_id = self.channel_ids[index];
-
-                let channel_join_request = mcs::ChannelJoinRequest {
-                    initiator_id: user_channel_id,
-                    channel_id,
-                };
+            // Send all the join requests in a single batch.
+            // > RDP 4.0, 5.0, 5.1, 5.2, 6.0, 6.1, 7.0, 7.1, 8.0, 10.2, 10.3,
+            // > 10.4, and 10.5 clients send a Channel Join Request to the server only after the
+            // > Channel Join Confirm for a previously sent request has been received. RDP 8.1,
+            // > 10.0, and 10.1 clients send all of the Channel Join Requests to the server in a
+            // > single batch to minimize the overall connection sequence time.
+            ChannelConnectionState::SendChannelJoinRequest { user_channel_id } => {
+                let mut written = 0;
+
+                for channel_id in self.channel_ids.iter().copied() {
+                    let channel_join_request = mcs::ChannelJoinRequest {
+                        initiator_id: user_channel_id,
+                        channel_id,
+                    };
 
-                debug!(message = ?channel_join_request, "Send");
+                    debug!(message = ?channel_join_request, "Send");
 
-                let written = ironrdp_pdu::encode_buf(&channel_join_request, output).map_err(ConnectorError::pdu)?;
+                    written += ironrdp_pdu::encode_buf(&channel_join_request, output).map_err(ConnectorError::pdu)?;
```

Comment:
written = written.checked_add(ironrdp_pdu::encode_buf(&channel_join_request, output).map_err(ConnectorError::pdu)?).unwrap();

Reply:
lgtm otherwise

Reply:
Thank you!